src/odemis/driver/xt_client.py

```diff
@@ -676,6 +684,131 @@ def __init__(self, name, role, parent, hfw_nomag, **kwargs):
+        est_start = time.time() + 0.1
+        f = ProgressiveFuture(start=est_start,
+                              end=est_start + 20)  # rough time estimation
+        f._auto_contrast_brighness_lock = threading.Lock()
+        f._must_stop = threading.Event()  # cancel of the current future requested
+        f.task_canceller = self._cancelAutoContrastBrightness
+        f._channel_name = DETECTOR2CHANNELNAME[detector]
+        return self._executor.submitf(f, self._applyAutoContrastBrightness, f)
+
+    def _applyAutoContrastBrightness(self, future):
+        """
+        Starts applying auto contrast brightness and checks if the process is finished for the ProgressiveFuture object.
+        :param future (Future): the future to start running.
+        """
+        channel_name = future._channel_name
+        with future._auto_contrast_brighness_lock:
+            if future._must_stop.is_set():
+                raise CancelledError()
+            self.parent.set_auto_contrast_brightness(channel_name, XT_RUN)
+            time.sleep(0.5)  # Wait for the autofocussing to start
+
+        # Wait until the microscope is no longer autofocussing
+        while self.parent.is_running_auto_contrast_brightness(channel_name):
+            future._must_stop.wait(0.1)
+            if future._must_stop.is_set():
+                raise CancelledError()
+
+    def _cancelAutoContrastBrightness(self, future):
+        """
+        Cancels the auto contrast brightness. Non-blocking.
+        :param future (Future): the future to stop.
+        :return (bool): True if it successfully cancelled (stopped) the move.
+        """
+        future._must_stop.set()  # tell the thread taking care of autofocussing it's over
+
+        with future._auto_contrast_brighness_lock:
+            logging.debug("Cancelling auto contrast brightness")
+            try:
+                self.parent.set_auto_contrast_brightness(future._channel_name, XT_STOP)
+                return True
+            except OSError as error_msg:
+                # Catch error when canceling fails because auto contrast brightness has already stopped.
+                if error_msg.find("Set auto contrast brightness state failed Error message: None") == -1:
+                    time.sleep(1.0)  # Give microscope/simulator time to update the state
+                    # Check if auto contrast brightness stopped
+                    if not self.parent.is_running_auto_contrast_brightness(future._channel_name):
+                        return False
+                    else:
+                        # Auto contrast brightness is not stopped correctly, then do so.
+                        time.sleep(2.0)  # Wait a bit so we can properly cancel auto contrast brightness.
+                        # Stop auto contrast brightness properly
+                        self.parent.set_auto_contrast_brightness(future._channel_name, XT_STOP)
+                        return True
```

Comment:
@pieleric Maybe this part is a bit over the top since we don't expect this to happen. But maybe because of threading the auto functionality is not started yet and it is already tried to cancel it. Then this error is raised but the functionality is not actually stopped. Very much a corner case and we could replace these lines with just "raise" but I don't know what you think is the nicest solution?